plugins/omazureeventhubs/omazureeventhubs.c

```diff
@@ -1459,27 +654,119 @@ CODESTARTtryResume
 	DBGPRINTF("omazureeventhubs: tryResume for pData %p returned %d\n", pData, iRet);
 ENDtryResume
 
+BEGINbeginTransaction
+CODESTARTbeginTransaction
+	/* we have nothing to do to begin a transaction */
+ENDbeginTransaction
+
 /* 
- *
+ *	New Transaction action interface
  */
+BEGINcommitTransaction
+	instanceData *__restrict__ const pData = pWrkrData->pData;
+	unsigned i;
+	sbool bDone = 0;
+	protonmsg_entry* pMsgEntry = NULL;
+CODESTARTcommitTransaction
+	pthread_mutex_lock(&pData->mut_doAction);
+#ifndef NDEBUG
+	DBGPRINTF("omazureeventhubs: commitTransaction for pData %p ENTER\n", pData);
+#endif
+	if (pData->bIsConnected == 0) {
+		CHKiRet(setupProtonHandle(pData, 0));
+	}
+
+	// Handle/Expand our proton helper array
+	pData->nProtonMsgs = 0;
+	if (nParams > pData->nMaxProtonMsgs) {
+		// Free old array first!
+		if(pData->aProtonMsgs != NULL) {
+			for(i = 0 ; i < pData->nProtonMsgs ; ++i) {
+				if (pData->aProtonMsgs[i] != NULL) {
+					protonmsg_entry_destruct(pData->aProtonMsgs[i]);
+				}
+			}
+			free(pData->aProtonMsgs);
+		}
+
+		// Expand our proton helper array
+		DBGPRINTF("omazureeventhubs: commitTransaction expand helper array from %d to %d for pData %p\n",
+			pData->nMaxProtonMsgs, nParams, pData);
+		pData->nMaxProtonMsgs = nParams; // Set new MAX
+		CHKmalloc(pData->aProtonMsgs = calloc(pData->nMaxProtonMsgs, sizeof(struct s_protonmsg_entry)));
```

Comment:
remalloc() is quicker. You may even keep the proton messages, depding on their exact semantics. But in case of remalloc() you must zero out the new elemets yourself.

Reply:
I chose calloc because of zeroing, and the message buffer expansion should not happen that often so from my point of view it is acceptable that calloc is a little less performant.

Reply:
It's actually more than you think. Because with calloc(), you need to free the previous array (else you get a memory leak). With relloc(), this is not necessary, as the previsou array is allocated. The important thing to note is that the free/calloc path requires two calls to system memory allocator. This is pretty expensive, because most allocators use a global lock for there operation, ending all concurrency for other threads using the malloc system. So each call makes a big difference. I am telling from experience. Also, this is one that that usually is not corrected when not done right initially. And usually it should be pretty cheap in terms of effort to do it right up from the begining. Just my 2 cts...